File: dart-sdk/test/integration/call_flow_test.dart
import 'dart:io';
import 'dart:convert';
import 'package:http/http.dart' as http;

/// Vobiz Dart SDK - Full Call Flow Integration Test
/// Uses direct HTTP requests for call_uuid operations.
Future<void> main() async {
  final authId         = Platform.environment['VOBIZ_AUTH_ID']         ?? '';
  final authToken      = Platform.environment['VOBIZ_AUTH_TOKEN']       ?? '';
  final fromNumber     = Platform.environment['VOBIZ_FROM_NUMBER']      ?? '';
  final toNumber       = Platform.environment['VOBIZ_TO_NUMBER']        ?? '';
  final transferNumber = Platform.environment['VOBIZ_TRANSFER_NUMBER']  ?? '';

  if (authId.isEmpty || authToken.isEmpty || fromNumber.isEmpty || toNumber.isEmpty) {
    print('SKIP: Required env vars not set');
    exit(0);
  }

  const audioUrl    = 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3';
  const answerUrl   = 'https://internal-test-xml.vobiz.ai/answer';
  const hangupUrl   = 'https://internal-test-xml.vobiz.ai/hangup';
  const transferUrl = 'https://internal-test-xml.vobiz.ai/answer';
  const baseUrl     = 'https://api.vobiz.ai';

  final headers = {
    'X-Auth-ID':    authId,
    'X-Auth-Token': authToken,
    'Content-Type': 'application/json',
  };

  Future<Map<String, dynamic>> vobizRequest(String method, String path, [Map<String, dynamic>? body]) async {
    final uri  = Uri.parse('$baseUrl$path');
    final bodyStr = body != null ? jsonEncode(body) : '';
    http.Response resp;
    switch (method) {
      case 'POST':   resp = await http.post(uri,   headers: headers, body: bodyStr); break;
      case 'DELETE': resp = await http.delete(uri, headers: headers, body: bodyStr); break;
      default:       resp = await http.get(uri,    headers: headers);
    }
    print('  -> HTTP ${resp.statusCode}');
    try { return jsonDecode(resp.body) as Map<String, dynamic>; }
    catch (_) { return {}; }
  }

  final sleep = (int sec) => Future.delayed(Duration(seconds: sec));
  int passed = 0, failed = 0;

  Future<void> step(String name, Future<void> Function() fn) async {
    try { await fn(); print('[Dart] PASS: $name'); passed++; }
    catch (e) { print('[Dart] FAIL: $name - $e'); failed++; }
  }

  // STEP 1: Make outbound call
  print('\n[Dart] STEP 1: Making outbound call...');
  String requestUUID = '';
  await step('Make Call', () async {
    final data = await vobizRequest('POST', '/api/v1/Account/$authId/Call/', {
      'from': fromNumber, 'to': toNumber,
      'answer_url': answerUrl, 'answer_method': 'POST',
      'hangup_url': hangupUrl, 'hangup_method': 'POST',
    });
    requestUUID = data['request_uuid'] as String? ??
        ((data['objects'] as List?)?.first as Map?)?.get('request_uuid') as String? ?? '';
    if (requestUUID.isEmpty) throw Exception('No request_uuid in: $data');
    print('  -> request_uuid = $requestUUID');
  });
  if (requestUUID.isEmpty) { print('[Dart] Cannot continue'); exit(1); }
  await sleep(5);

  await step('List Live Calls', () async {
    await vobizRequest('GET', '/api/v1/Account/$authId/Call/?status=live');
  });
  await sleep(5);

  await step('Get Single Live Call', () async {
    await vobizRequest('GET', '/api/v1/Account/$authId/Call/$requestUUID/?status=live');
  });
  await sleep(5);

  await step('Speak TTS', () async {
    await vobizRequest('POST', '/api/v1/Account/$authId/Call/$requestUUID/Speak/',
        {'text': 'Hello from Vobiz Dart SDK.', 'voice': 'WOMAN', 'language': 'en-US', 'legs': 'aleg'});
  });
  await sleep(5);

  await step('Stop TTS', () async {
    await vobizRequest('DELETE', '/api/v1/Account/$authId/Call/$requestUUID/Speak/');
  });

  await step('Play Audio', () async {
    await vobizRequest('POST', '/api/v1/Account/$authId/Call/$requestUUID/Play/',
        {'urls': [audioUrl], 'legs': 'aleg', 'loop': false, 'mix': true});
  });
  await sleep(5);

  await step('Stop Audio', () async {
    await vobizRequest('DELETE', '/api/v1/Account/$authId/Call/$requestUUID/Play/');
  });

  await step('Start Recording', () async {
    await vobizRequest('POST', '/api/v1/Account/$authId/Call/$requestUUID/Record/',
        {'time_limit': 60, 'file_format': 'mp3'});
  });
  await sleep(5);

  await step('Send DTMF', () async {
    await vobizRequest('POST', '/api/v1/Account/$authId/Call/$requestUUID/DTMF/',
        {'digits': '1234', 'leg': 'aleg'});
  });

  await step('Stop Recording', () async {
    await vobizRequest('DELETE', '/api/v1/Account/$authId/Call/$requestUUID/Record/');
  });

  await step('Transfer Call', () async {
    final transferTo = transferUrl + (transferNumber.isNotEmpty ? '?to=$transferNumber' : '');
    await vobizRequest('POST', '/api/v1/Account/$authId/Call/$requestUUID/',
        {'legs': 'aleg', 'aleg_url': transferTo, 'aleg_method': 'POST'});
  });
  await sleep(5);

  await step('Hang Up', () async {
    await vobizRequest('DELETE', '/api/v1/Account/$authId/Call/$requestUUID/');
  });

  print('\n[Dart] Call flow COMPLETE: $passed passed, $failed failed');
  exit(failed > 0 ? 1 : 0);
}

File: kotlin-sdk/src/test/kotlin/ai/vobiz/integration/CallFlowTest.kt
package ai.vobiz.integration

import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import org.json.JSONArray

/**
 * Vobiz Kotlin SDK - Full Call Flow Integration Test
 * Uses direct OkHttp calls for call_uuid operations.
 */
fun main() {
    val authId         = System.getenv("VOBIZ_AUTH_ID")         ?: ""
    val authToken      = System.getenv("VOBIZ_AUTH_TOKEN")      ?: ""
    val fromNumber     = System.getenv("VOBIZ_FROM_NUMBER")     ?: ""
    val toNumber       = System.getenv("VOBIZ_TO_NUMBER")       ?: ""
    val transferNumber = System.getenv("VOBIZ_TRANSFER_NUMBER") ?: ""

    if (authId.isEmpty() || authToken.isEmpty() || fromNumber.isEmpty() || toNumber.isEmpty()) {
        println("SKIP: VOBIZ_AUTH_ID, VOBIZ_AUTH_TOKEN, VOBIZ_FROM_NUMBER, VOBIZ_TO_NUMBER required")
        return
    }

    val AUDIO_URL    = "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3"
    val ANSWER_URL   = "https://internal-test-xml.vobiz.ai/answer"
    val HANGUP_URL   = "https://internal-test-xml.vobiz.ai/hangup"
    val TRANSFER_URL = "https://internal-test-xml.vobiz.ai/answer"
    val BASE_URL     = "https://api.vobiz.ai"
    val JSON_TYPE    = "application/json".toMediaType()

    val http = OkHttpClient()

    fun vobizRequest(method: String, path: String, body: Map<String, Any?>? = null): Pair<Int, JSONObject> {
        val payload = if (body != null) JSONObject(body).toString() else ""
        val rb: RequestBody? = when (method) {
            "POST"   -> payload.toRequestBody(JSON_TYPE)
            "DELETE" -> "".toRequestBody(JSON_TYPE)
            else     -> null
        }
        val req = Request.Builder()
            .url(BASE_URL + path)
            .addHeader("X-Auth-ID", authId)
            .addHeader("X-Auth-Token", authToken)
            .addHeader("Content-Type", "application/json")
            .apply {
                when (method) {
                    "GET"    -> get()
                    "POST"   -> post(rb!!)
                    "DELETE" -> delete(rb!!)
                }
            }.build()

        http.newCall(req).execute().use { resp ->
            val respBody = resp.body?.string() ?: "{}"
            val json = try { JSONObject(respBody) } catch (e: Exception) { JSONObject() }
            return Pair(resp.code, json)
        }
    }

    fun sleep(sec: Long) = Thread.sleep(sec * 1000)

    var passed = 0
    var failed = 0

    fun step(name: String, block: () -> Unit) {
        try { block(); println("[Kotlin] PASS: $name"); passed++ }
        catch (e: Exception) { println("[Kotlin] FAIL: $name - ${e.message}"); failed++ }
    }

    // STEP 1: Make outbound call
    println("\n[Kotlin] STEP 1: Making outbound call...")
    var requestUUID = ""
    step("Make Call") {
        val (status, data) = vobizRequest("POST", "/api/v1/Account/$authId/Call/", mapOf(
            "from" to fromNumber, "to" to toNumber,
            "answer_url" to ANSWER_URL, "answer_method" to "POST",
            "hangup_url" to HANGUP_URL, "hangup_method" to "POST"
        ))
        requestUUID = data.optString("request_uuid", "")
        if (requestUUID.isEmpty()) {
            val objects = data.optJSONArray("objects")
            if (objects != null && objects.length() > 0)
                requestUUID = objects.getJSONObject(0).optString("request_uuid", "")
        }
        if (requestUUID.isEmpty()) throw Exception("No request_uuid in: $data")
        println("  -> HTTP $status, request_uuid = $requestUUID")
    }
    if (requestUUID.isEmpty()) { println("[Kotlin] Cannot continue"); return }
    sleep(5)

    step("List Live Calls") {
        val (s, _) = vobizRequest("GET", "/api/v1/Account/$authId/Call/?status=live")
        println("  -> HTTP $s")
    }
    sleep(5)

    step("Get Single Live Call") {
        val (s, _) = vobizRequest("GET", "/api/v1/Account/$authId/Call/$requestUUID/?status=live")
        println("  -> HTTP $s")
    }
    sleep(5)

    step("Speak TTS") {
        val (s, _) = vobizRequest("POST", "/api/v1/Account/$authId/Call/$requestUUID/Speak/",
            mapOf("text" to "Hello from Vobiz Kotlin SDK.", "voice" to "WOMAN", "language" to "en-US", "legs" to "aleg"))
        println("  -> HTTP $s")
    }
    sleep(5)

    step("Stop TTS") {
        val (s, _) = vobizRequest("DELETE", "/api/v1/Account/$authId/Call/$requestUUID/Speak/")
        println("  -> HTTP $s")
    }

    step("Play Audio") {
        val (s, _) = vobizRequest("POST", "/api/v1/Account/$authId/Call/$requestUUID/Play/",
            mapOf("urls" to listOf(AUDIO_URL), "legs" to "aleg", "loop" to false, "mix" to true))
        println("  -> HTTP $s")
    }
    sleep(5)

    step("Stop Audio") {
        val (s, _) = vobizRequest("DELETE", "/api/v1/Account/$authId/Call/$requestUUID/Play/")
        println("  -> HTTP $s")
    }

    step("Start Recording") {
        val (s, _) = vobizRequest("POST", "/api/v1/Account/$authId/Call/$requestUUID/Record/",
            mapOf("time_limit" to 60, "file_format" to "mp3"))
        println("  -> HTTP $s")
    }
    sleep(5)

    step("Send DTMF") {
        val (s, _) = vobizRequest("POST", "/api/v1/Account/$authId/Call/$requestUUID/DTMF/",
            mapOf("digits" to "1234", "leg" to "aleg"))
        println("  -> HTTP $s")
    }

    step("Stop Recording") {
        val (s, _) = vobizRequest("DELETE", "/api/v1/Account/$authId/Call/$requestUUID/Record/")
        println("  -> HTTP $s")
    }

    step("Transfer Call") {
        val transferTo = TRANSFER_URL + if (transferNumber.isNotEmpty()) "?to=$transferNumber" else ""
        val (s, _) = vobizRequest("POST", "/api/v1/Account/$authId/Call/$requestUUID/",
            mapOf("legs" to "aleg", "aleg_url" to transferTo, "aleg_method" to "POST"))
        println("  -> HTTP $s")
    }
    sleep(5)

    step("Hang Up") {
        val (s, _) = vobizRequest("DELETE", "/api/v1/Account/$authId/Call/$requestUUID/")
        println("  -> HTTP $s")
    }

    println("\n[Kotlin] Call flow COMPLETE: $passed passed, $failed failed")
    if (failed > 0) System.exit(1)
}
